### DISCUSSION TITLE
Working with guards + hover while using @xstate/react? (loom and codesandbox reproduction included)

### DISCUSSION THREAD

**Discussion by mcabrams** [statelyai/xstate#1671] [Q&A] [answered]
Hi,

I'm getting a better feel for xstate and I am just beginning to use the @xstate/react package to try to implement some simple statecharts.  One little project I'm working on is using xstate for storing some of the state, with other state either residing in useState hooks or other places.

I've reconstructed a codesandbox (https://codesandbox.io/s/pensive-darwin-ew7ov) and loom going over my issue (https://www.loom.com/share/9bf742b2903746b3836e83b95209e1ab).  The basic idea is that when a user hovers over the list of numbers they should see a prompt to add a number, at which point they can click that and add a static number.  The `isHoveringGuard` is supposed to handle the scenario where the person has confirmed they want to add the number (at which point the entered menu should disappear and either return to the unhovered hidden state or the hovered state - I have tried to achieve this with the guard but for whatever reason the guard always returns false, any ideas on how to solve this?).  

This is a bit of a contrived example to reproduce this in codesandbox - I am aware I could use certain css for the hovering/focus functionality, but I'm also trying to learn how to use xstate so would appreciate to at least understand where my mistake is, thanks!

**Comment by Andarist** [comment, marked as answer]
It happens because we dont properly update guards so u suffer from the stale closure problem. Im going to fix this soon

**Reply by mcabrams** [reply]
lol I thought it might be closure related, it was the only thing I could think of especially since the guard always returned false.  Out of curiosity, how is this going to be solved and is there any workaround right now?

**Reply by Andarist** [reply]
It's just a matter of providing a similar updater effect for guards:
https://github.com/davidkpiano/xstate/blob/80ea4ba385a589d471aa355818be461d288c063a/packages/xstate-react/src/useMachine.ts#L244-L246

I have already the change locally but need to write a test for it - which I'll have to do tomorrow as I'm super tired already.

As to the workaround - this should work: https://codesandbox.io/s/cocky-noether-un4j4?file=/src/App.js . However, it seems that you have not provided `ref` from `useHover` to the appropriate element so you should fix that first.

**Comment by davidkpiano** [comment]
On a separate/related note, guards are one of those things that should be part of the machine definition and _not_ dynamic. Otherwise, the guard is not a pure function and your machine is no longer deterministic 😱 

How can you refactor this? Take a look at the dynamic guard:

```js
  const isHoveringGuard = () => {
    //     vvvvvvvvvv - this is coming from the outside (impure)
    return isHovering;
  };
```

What if you could make it so that the guard only relies on information inside the machine?


```js
  const isHoveringGuard = (context) => {
    //             vvvvvvvvvv - internal from context! (pure)
    return context.isHovering;
  };
```

You would need to send an event to the machine notifying it that the hovering state has changed:

```js
useEffect(() => {
  // TODO: implement the assign({ ... }) transition action for this
  // in the machine
  send({ type: 'HOVER_CHANGED', value: isHovering });
}, [isHovering]);
```

And then your `isHoveringGuard` no longer needs to be dynamic. Also, your machine becomes much more reusable in other components, and you isolate logic from any implementation details and prevent impure function calls.

**TL;DR: Guards should be pure functions.**

**Reply by mcabrams** [reply]
Thanks @davidkpiano - this is very helpful!
